// @neo4j/cypher-builder/dist/expressions/functions/scalar.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.nullIf = exports.character_length = exports.char_length = exports.valueType = exports.type = exports.toIntegerOrNull = exports.toInteger = exports.toFloatOrNull = exports.toFloat = exports.toBooleanOrNull = exports.toBoolean = exports.timestamp = exports.startNode = exports.size = exports.randomUUID = exports.properties = exports.length = exports.last = exports.id = exports.head = exports.endNode = exports.elementId = exports.coalesce = void 0;
const CypherFunctions_1 = require("./CypherFunctions");
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-coalesce)
 * @group Cypher Functions
 * @category Scalar
 */
function coalesce(expr, ...optionalExpr) {
    return new CypherFunctions_1.CypherFunction("coalesce", [expr, ...optionalExpr]);
}
exports.coalesce = coalesce;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-elementid)
 * @group Cypher Functions
 * @category Scalar
 */
function elementId(variable) {
    return new CypherFunctions_1.CypherFunction("elementId", [variable]);
}
exports.elementId = elementId;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-endnode)
 * @group Cypher Functions
 * @category Scalar
 */
function endNode(relationship) {
    return new CypherFunctions_1.CypherFunction("endNode", [relationship]);
}
exports.endNode = endNode;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-head)
 * @group Cypher Functions
 * @category Scalar
 */
function head(expr) {
    return new CypherFunctions_1.CypherFunction("head", [expr]);
}
exports.head = head;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-id)
 * @group Cypher Functions
 * @category Scalar
 * @deprecated Use {@link elementId} instead
 */
function id(variable) {
    return new CypherFunctions_1.CypherFunction("id", [variable]);
}
exports.id = id;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-last)
 * @group Cypher Functions
 * @category Scalar
 */
function last(expr) {
    return new CypherFunctions_1.CypherFunction("last", [expr]);
}
exports.last = last;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-length)
 * @group Cypher Functions
 * @category Scalar
 */
function length(path) {
    return new CypherFunctions_1.CypherFunction("length", [path]);
}
exports.length = length;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-properties)
 * @group Cypher Functions
 * @category Scalar
 */
function properties(expr) {
    return new CypherFunctions_1.CypherFunction("properties", [expr]);
}
exports.properties = properties;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-randomuuid)
 * @group Cypher Functions
 * @category Scalar
 */
function randomUUID() {
    return new CypherFunctions_1.CypherFunction("randomUUID");
}
exports.randomUUID = randomUUID;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-size)
 * @group Cypher Functions
 * @category Scalar
 */
function size(expr) {
    return new CypherFunctions_1.CypherFunction("size", [expr]);
}
exports.size = size;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-startnode)
 * @group Cypher Functions
 * @category Scalar
 */
function startNode(relationship) {
    return new CypherFunctions_1.CypherFunction("startNode", [relationship]);
}
exports.startNode = startNode;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-timestamp)
 * @group Cypher Functions
 * @category Scalar
 */
function timestamp() {
    return new CypherFunctions_1.CypherFunction("timestamp");
}
exports.timestamp = timestamp;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-toboolean)
 * @group Cypher Functions
 * @category Scalar
 */
function toBoolean(expr) {
    return new CypherFunctions_1.CypherFunction("toBoolean", [expr]);
}
exports.toBoolean = toBoolean;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-tobooleanornull)
 * @group Cypher Functions
 * @category Scalar
 */
function toBooleanOrNull(expr) {
    return new CypherFunctions_1.CypherFunction("toBooleanOrNull", [expr]);
}
exports.toBooleanOrNull = toBooleanOrNull;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-tofloat)
 * @group Cypher Functions
 * @category Scalar
 */
function toFloat(expr) {
    return new CypherFunctions_1.CypherFunction("toFloat", [expr]);
}
exports.toFloat = toFloat;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-tofloatornull)
 * @group Cypher Functions
 * @category Scalar
 */
function toFloatOrNull(expr) {
    return new CypherFunctions_1.CypherFunction("toFloatOrNull", [expr]);
}
exports.toFloatOrNull = toFloatOrNull;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-tointeger)
 * @group Cypher Functions
 * @category Scalar
 */
function toInteger(expr) {
    return new CypherFunctions_1.CypherFunction("toInteger", [expr]);
}
exports.toInteger = toInteger;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-tointegerornull)
 * @group Cypher Functions
 * @category Scalar
 */
function toIntegerOrNull(expr) {
    return new CypherFunctions_1.CypherFunction("toIntegerOrNull", [expr]);
}
exports.toIntegerOrNull = toIntegerOrNull;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-type)
 * @group Cypher Functions
 * @category Scalar
 */
function type(relationship) {
    return new CypherFunctions_1.CypherFunction("type", [relationship]);
}
exports.type = type;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-valueType)
 * @group Cypher Functions
 * @category Scalar
 */
function valueType(expr) {
    return new CypherFunctions_1.CypherFunction("valueType", [expr]);
}
exports.valueType = valueType;
/** Alias of size()
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-char_length)
 * @group Cypher Functions
 * @category Scalar
 */
function char_length(expr) {
    return new CypherFunctions_1.CypherFunction("char_length", [expr]);
}
exports.char_length = char_length;
/** Alias of size()
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-character_length)
 * @group Cypher Functions
 * @category Scalar
 * @version Neo4j 5.13
 */
function character_length(expr) {
    return new CypherFunctions_1.CypherFunction("character_length", [expr]);
}
exports.character_length = character_length;
/**
 * @see [Cypher Documentation](https://neo4j.com/docs/cypher-manual/current/functions/scalar/#functions-nullIf)
 * @group Cypher Functions
 * @category Scalar
 * @version Neo4j 5.14
 */
function nullIf(expr1, expr2) {
    return new CypherFunctions_1.CypherFunction("nullIf", [expr1, expr2]);
}
exports.nullIf = nullIf;
